mitiq/lre/inference/multivariate_richardson.py

```diff
@@ -0,0 +1,228 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+
+"""Functions for multivariate richardson extrapolation as defined in
+:cite:`Russo_2024_LRE`.
+"""
+
+import warnings
+from collections import Counter
+from itertools import combinations_with_replacement
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+from cirq import Circuit
+from numpy.typing import NDArray
+from sympy import Symbol
+
+from mitiq.lre.multivariate_scaling.layerwise_folding import (
+    _get_scale_factor_vectors,
+)
+
+
+def _full_monomial_basis_term_exponents(
+    num_layers: int, degree: int
+) -> List[Dict[int, int]]:
+    """Exponents of monomial terms required to create the sample matrix."""
+    variables = [i for i in range(1, num_layers + 1)]
+    variable_combinations = []
+    for d in range(degree, -1, -1):
+        for var_tuple in combinations_with_replacement(variables, d):
+            variable_combinations.append(var_tuple)
+
+    variable_exp_counter = [
+        dict(Counter(term)) for term in variable_combinations
+    ]
+
+    for combo_key in variable_exp_counter:
+        for j in variables:
+            if j not in combo_key:
+                combo_key[j] = 0
+
+    return variable_exp_counter[::-1]
+
+
+def full_monomial_basis_terms(num_layers: int, degree: int) -> List[str]:
+    r"""Find the monomial basis terms for a number of layers in the input
+    and max degree d.
+
+    Number of layers in the input circuit dictate the number of variables
+    utilized in the combinations used for the polynomial.
+
+    Degree of the polynomial dictates the max and min degree of the monomial
+    terms.
+
+    Args:
+        num_layers: Number of layers in the input circuit.
+        degree: Degree of the multivariate polynomial.
+
+    Returns:
+        Monomial basis terms required for multivariate
+            extrapolation up to max degree
+    """
```

Comment:
Note: This function is not used to evaluate the elements of the sample matrix. This is now a helper function for someone who wants to know which monomial terms are used to define the rows of the sample matrix.